### PR TITLE
Functoria

### PR DESCRIPTION
**Pull request by Drup**
This pull-request will, at term, replace all the internal of the mirage DSL by functoria: https://github.com/Drup/Functoria
It will also re-express all the combinators in the new DSL.

This new version of the DSL improves the following points:

- The module application tree is fully explicit, which make it explorable and printable at configuration time.
- It is possible to give parameters to the unikernel at configure, run time, or both and generate the configuration for all that.
- Sharing should be easier to express.
- The internal DSL to describe configurable devices should be easier to use (even if it uses classes *gasp*).
- Since mirage and the DSL are now separated, it's guaranteed to be extendable externally.

For now, I translated very few configurables.

What works now: the console example, with a key.
What doesn't work: all the rest.

### CODE REVIEW

**Comment by samoht** [comment]
Thanks! My only concern (for now) it that I **really** don't like that name...

**Comment by dbuenzli** [comment]
> Thanks! My only concern (for now) it that I really don't like that name...

What about at least dropping the `h`s. (and maybe suffix a further `lalalilala`).

**Comment by Drup** [comment]
> Thanks! My only concern (for now) it that I really don't like that name...

Ahah, I expected that. I'm still open to propositions. :p 
Note that the end user should never have to see it, it will just be a one of the mirage dependency.

**Comment by samoht** [comment]
Right, after a 30 seconds very intense brainstorming, what about:
- Funtore (to say with the Italian accent(
- Funcat (because that's a functor in the category of endomorphisms)

**Comment by avsm** [comment]
+1 for funcat!

**Comment by Drup** [comment]
The only associate I have for funcat is [this](https://www.youtube.com/watch?v=QH2-TGUlwu4).

And I'm not sure it's a good thing.

**Comment by Drup** [comment]
It will be renamed to "Functoria" (proposed by @dsheets )

**Comment by avsm** [comment]
+1

**Comment by samoht** [comment]
(and we are recording the "metafunctoria" 80's sound-track to be ready when we'll have a proper integration with meta-OCaml)

**Comment by avsm** [comment]
Listening to this while I wait! https://www.youtube.com/watch?v=9jK-NcRmVcw

**Comment by Drup** [comment]
Done.

[This is relevant](https://i.imgur.com/Ys7OyIF.jpg). :p

**Comment by Drup** [comment]
I'm in the middle of converting the devices. 
Just for the sake of unfair comparison: [before](https://github.com/mirage/mirage/blob/cd92d39eefda34e73fcd0d365afc6b5be62e78bf/lib/mirage.ml#L1342-L1407) -> [after](https://github.com/Drup/mirage/blob/db8202bad0abb257d36cb17cd6fdc53b55258a91/lib/mirage.ml#L811-L828).

**Comment by samoht** [comment]
Yay! this looks very nice (and short) :-)

**Comment by dsheets** [comment]
`dummy_conf` -> `zero_conf`/`empty_conf`/`null_conf`?

**Comment by mor1** [comment]
...or `template_conf` or `blank_conf`?

`zero_conf` is a bit overloaded perhaps.

On 19 August 2015 at 11:24, David Sheets <notifications@github.com> wrote:

> dummy_conf -> zero_conf/empty_conf/null_conf?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mirage/mirage/pull/441#issuecomment-132506819>.
>



-- 
Richard Mortier
mort@cantab.net

**Comment by samoht** [comment]
would be nice if you could refresh my memories on how to use objects as well at one point :p what's the difference between `method` and `method!` ... and why the `ip` and `tcp` argument share the same `'a` parameter?

**Comment by dsheets** [comment]
`_Co_pyright`?

File: myocamlbuild.ml [hunk 1]
```diff
@@ -621,4 +621,9 @@ let dispatch_default = MyOCamlbuildBase.dispatch_default conf package_default;;
 
 # 623 "myocamlbuild.ml"
 (* OASIS_STOP *)
-Ocamlbuild_plugin.dispatch dispatch_default;;
+
+open Ocamlbuild_plugin
+let () =
+  pflag ["ocaml"; "link"] "dontlink" (fun pkg -> S[A"-dontlink"; A pkg]) ;;
```
Comment (samoht) [inline myocamlbuild.ml]: why do you need this?

**Comment by Drup** [comment]
> dummy_conf -> zero_conf/empty_conf/null_conf?

Oh, I was just thinking that I should rename it to `base_conf`.

> what's the difference between method and method!

"!" is explicit overloading (same than in "open!").



> and why the ip and tcp argument share the same 'a parameter?

That has nothing to do with objects, it's just the versioning of ip. :)

File: myocamlbuild.ml (hunk 1 above)
Comment (Drup) [inline myocamlbuild.ml]: I don't need it really, it's just to ensure that the runtime is not unix-dependent, in order to stay mirage-compatible.

**Comment by samoht** [comment]
> "!" is explicit overloading (same than in open!).

yea, that's about all what I know about that construct (its name :-)). I don't know when to use it or not though (a quick read of `mirage.ml` didn't let me infer a clear rule.

Also, to be picky, in the example that you show there is no clear win at using objects. I am fine to use objects and class internally if that make the code simpler, but would it be possible to not expose it in the interface? I haven't looked at the details so I don't know if that's even possible.

**Comment by Drup** [comment]
Well, I can tell just by how the combinators are done how much some part of the stacks are figured out. :)

In the conduit combinator, first class modules are used to do some fancy initialization thing (`Conduit_mirage.with_tcp`). Can it be avoided (and transformed into a normal functor) ? It's completely unlike what the rest of the stack is doing.

In the direct resolver, two functors are applied. It looks like that:
"module %s = Resolver_lwt" (module_name t);
```ocaml
module Foo_dns = Dns_resolver_mirage.Make(OS.Time)(MyStack)
module Foo = Resolver_mirage.Make(Foo_dns)
```
I explicitly forbid this kind of things, so I have two choices: I add intermediate combinators or I add the combined functors in `conduit`. Which one should it be ?

In general, far too much things are hardcoded in these two combinators and the code should be moved to the respective libraries.

**Comment by samoht** [comment]
> In general, far too much things are hardcoded in these combinators and the code should be moved to the respective libraries.

I fully agree. Please feel free to create patches doing this (or create issues so that people who want to help can contribute).

>  explicitly forbid this kind of things, so I have two choices: I add intermediate combinators or I add the combined functor in conduit. Which one should it be ?

move things in `conduit`. I'd rather simplify the DSL internal and make the library more uniform than adding more complexity in the DSL. Also it makes things easier for the ones who don't use the DSL.

> first class modules are used to do some fancy initialization thing (Conduit_mirage.with_tcp). Can it be avoided (and transformed into a normal functor) ? It's completely unlike what the rest of the stack is doing.

This is indeed quite different from the rest. The annoying thing with conduit is that (i) it potentially has (or will have) a lot of optional compilation options (depending on whether you want tcp, tls and vchan) and (ii) you want to extend it easily without having to add dummy implementation for the implementation you don't want to depend on. See how it was before for instance: 

https://github.com/mirage/ocaml-conduit/blob/v0.7.2/lib/conduit_mirage.ml#L143

Using first-class modules you can pass a list of backend implementation instead of having a variable-arity functor (depending on your pre-processing options).

To be honnest I am not a huge fan of the current design of conduit as well, but I find it better than the previous one.

**Comment by Drup** [comment]
The issue with this initialization scheme is that I kept the functor application cleanly separated from the init part in the DSL: The connect code is only aware of 1) the current module name 2) the connect names from the dependencies. This is a nice property that I would like to preserve.
This scheme breaks it (by making the conduit connect code depending on the stackv4 module).

I will try to think of something that will not introduce arbitrary access in the functor application tree.

**Comment by samoht** [comment]
The new code looks simpler, thanks! This will be good :-)

**Comment by Drup** [comment]
mirage-www runs now, [here is the patch needed](https://github.com/Drup/mirage-www/commit/b12c42bd7fd2ccef668dfeb66d38e5531c45277f).

Now that I have reestablished most of the functionality with almost no change to the interface, I will 
- [ ] Extract things from the mirage DSL back to various libraries.
- [ ] Add keys everywhere
- [ ] Document functoria
- [ ] Attend ICFP.

I'm considering removing all of lib_test and transferring everything to mirage-skeleton. Half of them are shared and it would ensure uptodate-ness of both...

Currently, I have one issue bugging me. When a device connects to other devices, their connects are run *sequentially* through a code that look like that:

```ocaml 
let arpv41 () =
  ethif1 () >>= function
  | `Error _e -> fail (Failure "ethif1")
  | `Ok _ethif1 ->
  clock () >>= function
  | `Error _e -> fail (Failure "clock")
  | `Ok _clock ->
  time () >>= function
  | `Error _e -> fail (Failure "time")
  | `Ok _time ->
  F3.connect _ethif1
```

However, now, "mirage" is a configurable like any other, and it basically connects to all the jobs. Which means I need the connects to run *concurrently*. Does someone got the right piece of code to do that ? `Lwt.join` only works on unit thread.

**Comment by Drup** [comment]
And also: [behold this majestic functor application](https://github.com/Drup/mirage/blob/e2dd4d411644760c917f89bc1665b1a5020651c6/lib/mirage.ml#L834-L838).

**Comment by samoht** [comment]
Yay! Congrats :-)

> I'm considering removing all of lib_test and transferring everything to mirage-skeleton. Half of them are shared and it would ensure uptodate-ness of both...

yes remove that. Could be replaced that with proper unit-tests later if that make sense (I have no idea how to test that DSL though...)

> However, now, "mirage" is a configurable like any other, and it basically connects to all the jobs. Which means I need the connects to run concurrently. Does someone got the right piece of code to do that ? Lwt.join only works on unit thread.

It there is no data-dependencies between the connects then it's perfectly fine (and better!) to run them concurrently.

**Comment by Drup** [comment]
Data-dependencies should be modeled by the new `dependencies` method in all configurables, so that should not be an issue. 

The problem is that I simply can't figure out the good way to do it.

**Comment by samoht** [comment]
`Lwt_list.map_p`? :-)

**Comment by Drup** [comment]
I wanted to avoid the whole "let's build a list and then do a partial pattern match against it".

**Comment by talex5** [comment]
@Drup I may be misunderstanding the problem, but the concurrent version of your example is just:

    let arpv41 () =
      let _ethif1 = ethif1 () in
      let _clock = clock () in
      let _time = time () in
      _ethif1 >>= function
      | `Error _e -> fail (Failure "ethif1")
      | `Ok _ethif1 ->
      _clock >>= function
      | `Error _e -> fail (Failure "clock")
      | `Ok _clock ->
      _time >>= function
      | `Error _e -> fail (Failure "time")
      | `Ok _time ->
      F3.connect _ethif1 

i.e. start all the threads in sequence, then wait for them to finish in sequence.

(however, this example doesn't make much sense, because `_clock` and `_time` are unused)

**Comment by Drup** [comment]
@talex5 I ... didn't though that would work. That's exactly what I wanted.

**Comment by Drup** [comment]
I would like a review/opinion on [the block section](https://github.com/Drup/mirage/blob/9257f1a6038fb5dd0007d7a7afad480a4b634da6/lib/mirage.ml#L272-L410). 

The way I do things now is that I have only one combinator for fat and I added combinators that creates a block with fat things in it. Same things from archive/tar. This way, `fat_from_files` is simply a composition of the two. 

cc @djs55 as you are the original author.

**Comment by Drup** [comment]
List of related PRs:
- [ ] https://github.com/mirage/ocaml-conduit/pull/92
- [ ] https://github.com/mirage/mirage-tcpip/pull/170
- [ ] https://github.com/mirage/mirage-http/pull/20

**Comment by Drup** [comment]
I have made various progress on the key front and the documentation front. Here are some screenshots:
https://i.imgur.com/wf2B4Mz.png
https://i.imgur.com/dN7t5XC.png

**Comment by Drup** [comment]
Here is the implementation of the generic stack: https://github.com/Drup/mirage/blob/612ae96bb6d5dec45617a4ee9aa101447b9adef6/lib/mirage.ml#L873-L883

(The argument `stack` allows to name the substack, which adds prefix to the keys)

**Comment by samoht** [comment]
That's really great!!! I think we need to think of a better way to show the configuration tree - the sharing is indeed not showed and the tree is currently not very readable: but it's really useful to have it and be able to show it! I'll try to think about it a bit more and we can discuss about that tomorrow.

**Comment by samoht** [comment]
And the generic stack example is great as well :-) maybe rename `?stack` to `?name` to me more descriptive?

**Comment by mor1** [comment]
Looks really nice! Minor request-- indicating what to specify other than `None` to `--tracing=` would be useful... :)

**Comment by Drup** [comment]
I rewrote the engine (for the 6th time!) to something completely graph based. It might be a bit slower, but it doesn't matter for the small graph we have, and it's much nicer to work with!

Dot of the generic stack
green/red are the then/else branches of the If. Bold is the default. Squares are actual configurables.

**Comment by Drup** [comment]
Note: the graph present the exact same sharing that what will happen in the generated code (once the graph is simplified by the keys).

**Comment by samoht** [comment]
Wow! That graph is very clear, I like it :-)

**Comment by Drup** [comment]
I'm considering dropping the textual representation completely. It's quite hard to show anything with it. If people are interested in a textual representation: What do you want to see in it ? 

I tried transforming the dot in text, the output is [adorable](https://gist.github.com/Drup/9db87514f7b978cbde62) but might not scale very well.

**Comment by samoht** [comment]
To prepare the integration, would it be possible to:
- [ ] move the Functoria project into the mirage org. (as it is now a core project for MirageOS) Obviously, we will give you full push/pull rights on the repo (just need to add you to the mirage group if you are not already in there)
- [ ] update the breaking API changes page in mirage-www, to document what needs to be modified in existing `config.ml` (especially regarding the TLS combinators)
- [ ] prepare a blog post, some documentation about the changes and show how to define and use the new (awesome) configuration/runtime keys.

I'll look and merge the open PR tomorrow morning.

**Comment by avsm** [comment]
This is a pretty big breaking change for existing configs, and it's hard to migrate old to new I presume.  A bump to Mirage 3.0 seems in order...

**Comment by samoht** [comment]
@avsm all the old config.ml should continue to work. They will just not have the new keys.

**Comment by Drup** [comment]
> This is a pretty big breaking change for existing configs, and it's hard to migrate old to new I presume. A bump to Mirage 3.0 seems in order...

It shouldn't be. See first 2 commits in my PR to mirage-www.

**Comment by samoht** [comment]
The PR are now merged and the releases have been done to opam-repository (they are not yet merged):

- bootvar: https://github.com/ocaml/opam-repository/pull/4826
- tcpip: https://github.com/ocaml/opam-repository/pull/4827
- conduit: https://github.com/ocaml/opam-repository/pull/4828
- mirage-http: https://github.com/ocaml/opam-repository/pull/4829

@djs55 is looking into the (not urgent) fat scripts and xl generation libs. If you can rebase this branch, I'll be able to test it on mirage-skeleton and be more confident that it's mergeable :-)

**Comment by Drup** [comment]
I didn't tried a lot of things, so testing should be made before release. There are also lot's of details that should be discussed about the specific implementation of the combinators. At the moment, all the network combinators that need parameters takes keys and the exposed functions create a new key with a default and pass it to the implementation. This means, among other thing, that `tap0` or `stack_with_static_ip` have a key to change their internal option.

Basically: please read and test. :3

**Comment by Drup** [comment]
Also relevant: functoria uses fmt, which is not released (@dbueznli :)

**Comment by samoht** [comment]
All new releases are in opam now. Can you rebase the patch it so I can test it? thx

**Comment by Drup** [comment]
Merge done, I had to introduce some modification to make it work with keys. I use the guarantee, provided by functoria, that all the configure methods are run in reversed topological order, hence mirage's configure is run last. @djs55 please review. 

It works well with keys.

**Comment by Drup** [comment]
@samoht There is some weirdness in the travis script related to mirage-types that I don't understand, it tries to compiles mirage too ?

**Comment by Drup** [comment]
@samoht What's missing before merge ? I have a blog post mostly ready [here](https://github.com/Drup/mirage-www/tree/funtoria%2Bpost).

**Comment by samoht** [comment]
I think it's almost ready :p  

can you create a PR with the blog post to make it easy to review ? (/cc @amirmc)
The only two last items which are almost there are:

- if we have a way to not break `--xen` and `--unix` that would be awesome (even if it is an horrible hack). We have lots of material (such as screencasts) which are using that, so it's a bit bad to break it. Or at least, we should have a proper error message to tell the user what to use now

- support for env variable is now in cmldiner. If you have time to integrate that and check that it works fine, @dbuenzli can make a new release and we'll use it.

Appart from that, that's it! Just need @avsm to have a quick look and ack and I'm happy to merge -- minor regression can be fixed later.

**Comment by amirmc** [comment]
> can you create a PR with the blog post to make it easy to review ? (/cc @amirmc)

Yes, please.  A PR that has just the post would be ideal and I can comment using the GitHub interface.Or you can put it on a wiki page and we can both edit it there before you submit a PR.  That might be easier.

**Comment by samoht** [comment]
Why is the CI failing?

```
#=== ERROR while installing mirage.2.7.0 ======================================#
# opam-version 1.2.2
# os           linux
# command      make
# path         /home/travis/.opam/system/build/mirage.2.7.0
# compiler     system (4.02.3)
# exit-code    2
# env-file     /home/travis/.opam/system/build/mirage.2.7.0/mirage-11224-d2c37b.env
# stdout-file  /home/travis/.opam/system/build/mirage.2.7.0/mirage-11224-d2c37b.out
# stderr-file  /home/travis/.opam/system/build/mirage.2.7.0/mirage-11224-d2c37b.err
### stdout ###
# [...]
# /home/travis/.opam/system/bin/ocamlfind ocamlc -c -g -annot -bin-annot -safe-string -short-paths -principal -keep-locs -strict-sequence -I lib_runtime -I lib -package functoria -package functoria.runtime -package ipaddr -w +A-4-6-7-9-40-42-44-48 -I lib -I lib_runtime -o lib/mirage.cmi lib/mirage.mli
# /home/travis/.opam/system/bin/ocamlfind ocamldep -package functoria -package functoria.runtime -package ipaddr -modules lib/mirage.ml > lib/mirage.ml.depends
# /home/travis/.opam/system/bin/ocamlfind ocamldep -package functoria -package functoria.runtime -package ipaddr -modules lib/mirage_version.ml > lib/mirage_version.ml.depends
# /home/travis/.opam/system/bin/ocamlfind ocamlc -c -g -annot -bin-annot -safe-string -short-paths -principal -keep-locs -strict-sequence -I lib_runtime -I lib -package functoria -package functoria.runtime -package ipaddr -w +A-4-6-7-9-40-42-44-48 -I lib -I lib_runtime -o lib/mirage_version.cmo lib/mirage_version.ml
# /home/travis/.opam/system/bin/ocamlfind ocamldep -package functoria -package functoria.runtime -package ipaddr -modules lib/mirage_key.ml > lib/mirage_key.ml.depends
# /home/travis/.opam/system/bin/ocamlfind ocamlc -c -g -annot -bin-annot -safe-string -short-paths -principal -keep-locs -strict-sequence -I lib_runtime -I lib -package functoria -package functoria.runtime -package ipaddr -w +A-4-6-7-9-40-42-44-48 -I lib -I lib_runtime -o lib/mirage.cmo lib/mirage.ml
# + /home/travis/.opam/system/bin/ocamlfind ocamlc -c -g -annot -bin-annot -safe-string -short-paths -principal -keep-locs -strict-sequence -I lib_runtime -I lib -package functoria -package functoria.runtime -package ipaddr -w +A-4-6-7-9-40-42-44-48 -I lib -I lib_runtime -o lib/mirage.cmo lib/mirage.ml
# File "lib/mirage.ml", line 26, characters 12-23:
# Error: Unbound value Info.keymap
# Command exited with code 2.
```

**Comment by Drup** [comment]
I don't know, I think it tries to pick up the wrong version of functoria somehow.

**Comment by samoht** [comment]
Indeed, this should be fixed by https://github.com/mirage/mirage-dev/pull/100

**Comment by Drup** [comment]
@samoht There is still the issue with the travis test for mirage-types.

**Comment by samoht** [comment]
Which issue? Is it this one? 

```
#=== ERROR while installing mirage.2.7.0 ======================================#
# opam-version 1.2.2
# os           linux
# command      make
# path         /home/travis/.opam/system/build/mirage.2.7.0
# compiler     system (4.01.0)
# exit-code    2
# env-file     /home/travis/.opam/system/build/mirage.2.7.0/mirage-9601-384e1f.env
# stdout-file  /home/travis/.opam/system/build/mirage.2.7.0/mirage-9601-384e1f.out
# stderr-file  /home/travis/.opam/system/build/mirage.2.7.0/mirage-9601-384e1f.err
### stdout ###
# [...]
# /home/travis/.opam/system/bin/ocamlfind ocamldep -package functoria.runtime -package ipaddr -modules lib_runtime/mirage_runtime.mli > lib_runtime/mirage_runtime.mli.depends
# /home/travis/.opam/system/bin/ocamlfind ocamlc -c -g -annot -bin-annot -short-paths -principal -strict-sequence -package functoria.runtime -package ipaddr -w +A-4-6-7-9-40-42-44-48 -I lib_runtime -o lib_runtime/mirage_runtime.cmi lib_runtime/mirage_runtime.mli
# /home/travis/.opam/system/bin/ocamlfind ocamldep -package functoria.runtime -package ipaddr -modules lib_runtime/mirage_runtime.ml > lib_runtime/mirage_runtime.ml.depends
# /home/travis/.opam/system/bin/ocamlfind ocamlc -c -g -annot -bin-annot -short-paths -principal -strict-sequence -package functoria.runtime -package ipaddr -w +A-4-6-7-9-40-42-44-48 -I lib_runtime -o lib_runtime/mirage_runtime.cmo lib_runtime/mirage_runtime.ml
# + /home/travis/.opam/system/bin/ocamlfind ocamlc -c -g -annot -bin-annot -short-paths -principal -strict-sequence -package functoria.runtime -package ipaddr -w +A-4-6-7-9-40-42-44-48 -I lib_runtime -o lib_runtime/mirage_runtime.cmo lib_runtime/mirage_runtime.ml
# File "lib_runtime/mirage_runtime.ml", line 43, characters 29-35:
# Error: Signature mismatch:
#        ...
#        The field `pp_hum' is required but not provided
# Command exited with code 2.
```

**Comment by Drup** [comment]
@samoht Yes, It shouldn't even try to compile mirage, only mirage-types (which contains only .mli files ...)

**Comment by Drup** [comment]
(Let met just say that I disapprove of having two (independent!) packages in one code base, it's just asking for mess build scripts)

**Comment by samoht** [comment]
> It shouldn't even try to compile mirage

I don't understand, please clarify what packages need to be pinned and which one shouldn't, so we can try to fix the CI. Do you want to do it yourself on that PR or is it something than I can fix somewhere else?

**Comment by Drup** [comment]
@samoht Well, it says "PACKAGE=mirage-types" on this line. I don't know travis-skeleton very well, but why should the script try to compile mirage in this case ?

**Comment by samoht** [comment]
Because of line 6 or .travis-ci.sh :-)

the scripts will compile the asked package, and then run some tests to check for breakage. So maybe this is a missing constraint to mirage-type? I'll check.

**Comment by samoht** [comment]
Needs a constraint to ipadrr version. `Ipaddr.pp_hum`

**Comment by samoht** [comment]
I'll fix that in the repo.

**Comment by Drup** [comment]
@samoht https://github.com/mirage/mirage-dev/pull/99 ?

**Comment by samoht** [comment]
Ha yes, thanks! I've relaunched the failed again... hopefully I haven't forget anything this time.

**Comment by samoht** [comment]
All green!

@Drup Any opinion on `--xen` and `--unix`?

**Comment by Drup** [comment]
I'm thinking about it, it's not trivial, see https://github.com/mirage/functoria/issues/14

**Comment by samoht** [comment]
Also there are some minor fixes to do to the breaking API change page in https://github.com/mirage/mirage-www/pull/390

Release list [no owner for the items yet]:

- [ ] decide if we keep support for `--xen` and `--unix` or not
- [ ] release functoria from mirage-dev to opam-repo
- [ ] release mirage from mirage-dev to opam-repo
- [ ] merge mirage-www updates https://github.com/mirage/mirage-www/pull/389
- [ ] merge mirage-skeleton updates https://github.com/mirage/mirage-skeleton/pull/104
- [ ] merge API breaking change on mirage-www: https://github.com/mirage/mirage-www/pull/390
- [ ] merge blog post new functoria on mirage-www

**Comment by Drup** [comment]
Ok, I have `--unix`, `--xen` and the `MODE` env variable back now.

I'm not convinced I should add env variables for all the predefined keys (like `DHCP`, and so on).

**Comment by Drup** [comment]
@samoht I'm not sure how to do the unix/macos autodetection propertly at the moment, unfortunately.

**Comment by samoht** [comment]
CI needs https://github.com/mirage/functoria/pull/24

@Drup Thanks! What is your worry about env variable? What's the problem with auto-detection?

**Comment by samoht** [comment]
```
#=== ERROR while installing mirage.2.6.1 ======================================#
# opam-version 1.2.2
# os           linux
# command      make
# path         /home/travis/.opam/system/build/mirage.2.6.1
# compiler     system (4.01.0)
# exit-code    2
# env-file     /home/travis/.opam/system/build/mirage.2.6.1/mirage-10848-4f1270.env
# stdout-file  /home/travis/.opam/system/build/mirage.2.6.1/mirage-10848-4f1270.out
# stderr-file  /home/travis/.opam/system/build/mirage.2.6.1/mirage-10848-4f1270.err
### stdout ###
# [...]
# /home/travis/.opam/system/bin/ocamlfind ocamlopt -shared -dontlink unix lib_runtime/mirage-runtime.cmxa lib_runtime/mirage_runtime.cmx -o lib_runtime/mirage-runtime.cmxs
# /home/travis/.opam/system/bin/ocamlfind ocamldep -package functoria -package functoria.runtime -package ipaddr -modules lib/mirage.mli > lib/mirage.mli.depends
# /home/travis/.opam/system/bin/ocamlfind ocamldep -package functoria -package functoria.runtime -package ipaddr -modules lib/mirage_key.mli > lib/mirage_key.mli.depends
# /home/travis/.opam/system/bin/ocamlfind ocamlc -c -g -annot -bin-annot -short-paths -principal -strict-sequence -I lib_runtime -I lib -package functoria -package functoria.runtime -package ipaddr -w +A-4-6-7-9-40-42-44-48 -I lib -I lib_runtime -o lib/mirage_key.cmi lib/mirage_key.mli
# /home/travis/.opam/system/bin/ocamlfind ocamlc -c -g -annot -bin-annot -short-paths -principal -strict-sequence -I lib_runtime -I lib -package functoria -package functoria.runtime -package ipaddr -w +A-4-6-7-9-40-42-44-48 -I lib -I lib_runtime -o lib/mirage.cmi lib/mirage.mli
# /home/travis/.opam/system/bin/ocamlfind ocamldep -package functoria -package functoria.runtime -package ipaddr -modules lib/mirage.ml > lib/mirage.ml.depends
# + /home/travis/.opam/system/bin/ocamlfind ocamldep -package functoria -package functoria.runtime -package ipaddr -modules lib/mirage.ml > lib/mirage.ml.depends
# File "lib/mirage.ml", line 1571, characters 10-11:
# Error: Syntax error
# Command exited with code 2.
```

**Comment by samoht** [comment]
@Drup can't you re-use https://github.com/mirage/mirage/blob/master/lib/main.ml#L60-75 for platform auto-detection?

**Comment by samoht** [comment]
```
# File "src/functoria_key.ml", line 123, characters 24-35:
# Error: Unbound value Arg.env_var
```

@Drup Can you add a pin to cmdliner? @dbuenzli seems that your env variable support for cmdliner works fine, would it be possible to have a new release?

**Comment by samoht** [comment]
Closing as superseded by #450
